File: tests/core/separators/codac2_tests_SepPolytope.cpp
/** 
 *  Codac tests
 * ----------------------------------------------------------------------------
 *  \date       2024
 *  \author     Simon Rohou
 *  \copyright  Copyright 2024 Codac Team
 *  \license    GNU Lesser General Public License (LGPL)
 */

#include <catch2/catch_test_macros.hpp>
#include <codac2_SepPolytope.h>
#include <codac2_CtcPolytope.h>
#include <codac2_pave.h>
#include <codac2_Figure3D.h>

using namespace std;
using namespace codac2;

TEST_CASE("SepPolytope")
{
   std::vector<std::pair<Row,double>> facets 
       { { {1,1,1}, 3.0 },
         { {-1,1,1}, 3.0 },
         { {1,-1,1}, 3.0 },
         { {1,1,-1}, 3.0 },
         { {-1,-1,1}, 3.0 },
         { {-1,1,-1}, 3.0 },
         { {1,-1,-1}, 3.0 },
         { {-1,-1,-1}, 3.0 } };
  Polytope p({{-2,2.0},{-2.0,2.0},{-2.0,2.0}}, facets);
  SepPolytope s(p);

  IntervalVector x(3);
  auto xs = s.separate(x);
  CHECK(xs.outer == IntervalVector({{-2,2},{-2,2},{-2,2}}));
  CHECK(xs.inner == IntervalVector(3));

  x = IntervalVector({{1.02,2.0},{1.02,2.0},{1.02,2.0}}); 
  xs = s.separate(x);
  CHECK(xs.inner == x);
  CHECK(xs.outer.is_empty());

}

File: tests/core/separators/codac2_tests_SepPolytope.py
#!/usr/bin/env python

#  Codac tests
# ----------------------------------------------------------------------------
#  \date       2025
#  \author     Simon Rohou
#  \copyright  Copyright 2024 Codac Team
#  \license    GNU Lesser General Public License (LGPL)

import unittest
from codac import *

class TestSepPolytope(unittest.TestCase):

  def test_SepPolytope(self):

    # Polytope, defined as a set of facets
    facets =[ [ Row([1,1,1]), 3.0 ],
              [ Row([-1,1,1]), 3.0 ],
              [ Row([1,-1,1]), 3.0 ],
              [ Row([1,1,-1]), 3.0 ],
              [ Row([-1,-1,1]), 3.0 ],
              [ Row([-1,1,-1]), 3.0 ],
              [ Row([1,-1,-1]), 3.0 ],
              [ Row([-1,-1,-1]), 3.0 ] ]

    p = Polytope(IntervalVector([[-2,2],[-2,2],[-2,2]]),facets,True)
    s = SepPolytope(p)

    x = IntervalVector(3)
    inner,outer = s.separate(x)
    self.assertTrue(inner == IntervalVector(3))
    self.assertTrue(outer == IntervalVector([[-2,2],[-2,2],[-2,2]]))

    x = IntervalVector([[1.02,2.0],[1.02,2.0],[1.0,2.0]])
    inner,outer = s.separate(x)
    self.assertTrue(inner == x)
    self.assertTrue(outer.is_empty())

if __name__ ==  '__main__':
  unittest.main()
